packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -84,6 +92,113 @@ export default createRule({
             descriptor,
             uses: identifierCounts === 1 ? 'never used' : 'used only once',
           },
+          suggest: [
+            {
+              messageId: 'replaceUsagesWithConstraint',
+              *fix(fixer): Generator<TSESLint.RuleFix> {
+                // Replace all the usages of the type parameter with the constraint...
+
+                const constraint = esTypeParameter.constraint;
+                // special case - a constraint of 'any' actually acts like 'unknown'
+                const constraintText =
+                  constraint != null &&
+                  constraint.type !== AST_NODE_TYPES.TSAnyKeyword
+                    ? context.sourceCode.getText(constraint)
+                    : 'unknown';
+                for (const reference of smTypeParameterVariable.references) {
+                  if (reference.isTypeReference) {
+                    const referenceNode = reference.identifier;
+                    yield fixer.replaceText(referenceNode, constraintText);
+                  }
+                }
+
+                // ...and remove the type parameter itself from the declaration.
+
+                const typeParams = nullThrows(
+                  node.typeParameters,
+                  'node should have type parameters',
+                ).params;
+
+                const index = typeParams.indexOf(esTypeParameter);
+
+                /* istanbul ignore if: this is an assertion that should never happen */
+                if (index === -1) {
+                  throw new Error(
+                    "type parameter should be in node's type parameters",
+                  );
+                } else if (typeParams.length === 1) {
```

Comment:
Why include this, if it's not something that can even be unit tested?

```suggestion
                if (typeParams.length === 1) {
```

Reply:
My paranoia is: the rule fixer _assumes_ the type parameter to be removed/replaced is _in_ the method/function/class's declared list of type parameters of the node being inspected. But the way we get the type parameter to be removed is indirect. 
1. ESTree selector inspects function/method: https://github.com/typescript-eslint/typescript-eslint/blob/9a4274095907d652d2abf07ac4f733a4eea6c121/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L206-L226
2. We switch into the ts AST node world with a type assertion: https://github.com/typescript-eslint/typescript-eslint/blob/9a4274095907d652d2abf07ac4f733a4eea6c121/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L43-L46
3. We traverse to its type parameters and back out to the ESTree type parameter node https://github.com/typescript-eslint/typescript-eslint/blob/9a4274095907d652d2abf07ac4f733a4eea6c121/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L54-L58
4. We _assume_ that the ESTree node from step 1. is the _grandparent_ of the type parameter from 3, in order to get its type parameter declaration (which we have to assert even exists): https://github.com/typescript-eslint/typescript-eslint/blob/9a4274095907d652d2abf07ac4f733a4eea6c121/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L117-L120
5. We need to figure out what position the reported ESTree type parameter is within that asserted-to-exist type parameter declaration's list of type parameters. My goal was to _double-check_ here that the reported type parameter really _is_ in that list, since, we have come here by two different paths completely, and it doesn't seem that far-fetched that a change to the rule logic could break some of our assumptions.

For reference, the ASTs look like
```text
ESTREE: node that we inspect ---(maybe has) ---> TSTypeParameterDeclaration -----> Array<TypeParameter>
TS AST: node that we inspect -----> Array<TypeParameterDeclaration>
```

Reply:
Anyways, I can go ahead and remove it if you prefer. Just one of those things that I prefer to do so that we fail faster  and more understandably when a nonlocal code change breaks some of our assumptions.

Reply:
Yeah this kind of strategy used to be a lot more common in the project. But it practically never paid off 😞. Our test coverage is so high & users are so active in reporting that it just ended up being a lot of code bloat. I'm still a -1 on it, but if you want to advocate for it I wouldn't be opposed to hearing more inputs from the team.

Reply:
i chop 🔪